File: Concurrency/1115.交替打印FooBar/FooBar.java
class FooBar {
    private int n;
    private Semaphore fooSem = new Semaphore(0);
    private Semaphore barSem = new Semaphore(1);

    public FooBar(int n) {
        this.n = n;
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            barSem.acquire();
            // printFoo.run() outputs "foo". Do not change or remove this line.
            printFoo.run();
            fooSem.release();
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            fooSem.acquire();
            // printBar.run() outputs "bar". Do not change or remove this line.
            printBar.run();
            barSem.release();
        }
    }
}

File: Concurrency/1115/FooBar.py
from threading import Semaphore

class FooBar:
    def __init__(self, n):
        self.n = n
        self.foo_sem = Semaphore(0)
        self.bar_sem = Semaphore(1)

    def foo(self, printFoo: 'Callable[[], None]') -> None:
        for i in range(self.n):
            self.bar_sem.acquire()
            # printFoo() outputs "foo". Do not change or remove this line.
            printFoo()
            self.foo_sem.release()


    def bar(self, printBar: 'Callable[[], None]') -> None:
        for i in range(self.n):
            self.foo_sem.acquire()
            # printBar() outputs "bar". Do not change or remove this line.
            printBar()
            self.bar_sem.release()
